File: rugs_env.go
package main

import (
	"encoding/base64"
	"errors"
	"fmt"
	"html"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"reflect"
	"runtime"
	"strings"
	"time"

	xml2json "github.com/basgys/goxml2json"
	"github.com/robertkrimen/otto"
)

func bindRugFunction(f func(call otto.FunctionCall) otto.Value) string {
	name := runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
	name = strings.Replace(name, "main.", "__duder_", -1)
	Duder.Log(LogVerbose, "[Rugs.BindRugFunction] Binding", name)
	Duder.Rugs.VM.Set(name, f)
	return name
}

func createRugEnvironment() error {
	data, err := os.ReadFile("rugs_env.js")
	if err != nil {
		return errors.New("unable to read rugs_env.js")
	}
	content := strings.Replace(string(data), "__BIND__", "%s", -1)
	// create the rug environment
	env := fmt.Sprintf(content,
		/* Duder */
		bindRugFunction(rugenvSetStatus),
		bindRugFunction(rugenvSetAvatar),
		bindRugFunction(rugenvSaveAvatar),
		bindRugFunction(rugenvGetAvatars),
		bindRugFunction(rugenvUseAvatar),
		bindRugFunction(rugenvStartTyping),
		/* DuderUser */
		bindRugFunction(rugenvRugUserGetIsOwner),
		bindRugFunction(rugenvRugUserGetIsManager),
		bindRugFunction(rugenvRugUserGetIsModerator),
		bindRugFunction(rugenvRugUserSetNickname),
		bindRugFunction(rugenvRugUserGetUsernameByID),
		bindRugFunction(rugenvRugUserGetIDByNickname),
		/* DuderCommand */
		bindRugFunction(rugenvRugCommandReplyToChannel),
		bindRugFunction(rugenvRugCommandReplyToChannelEmbed),
		bindRugFunction(rugenvRugCommandReplyToAuthor),
		bindRugFunction(rugenvRugCommandDeleteMessage),
		bindRugFunction(rugenvRugCommandSendFile),
		/* DuderMessageReaction */
		bindRugFunction(rugenvRugCommandReplyToChannel),
		bindRugFunction(rugenvRugCommandReplyToChannelEmbed),
		/* DuderRug */
		bindRugFunction(rugenvRugCreate),
		bindRugFunction(rugenvRugAddCommand),
		bindRugFunction(rugenvRugBindOnMessage),
		bindRugFunction(rugenvRugBindOnMessageReactionAdd),
		bindRugFunction(rugenvRugBindOnMessageReactionRemove),
		bindRugFunction(rugenvRugBindOnPresenceUpdate),
		bindRugFunction(rugenvRugLoadStorage),
		bindRugFunction(rugenvRugSaveStorage),
		bindRugFunction(rugenvRugDPrint),
		bindRugFunction(rugenvRugWPrint),
		/* String */
		bindRugFunction(rugenvStringDecodeHTML),
		/* HTTP */
		bindRugFunction(rugenvHTTPGet),
		bindRugFunction(rugenvHTTPPost),
		bindRugFunction(rugenvHTTPDetectContentType),
		bindRugFunction(rugenvHTTPParseURL),
		/* Base64 */
		bindRugFunction(rugenvBase64EncodeToString),
		/* XML */
		bindRugFunction(rugenvXMLtoJSON))

	if _, err := Duder.Rugs.VM.Run(env); err != nil {
		//fmt.Print(env)
		return errors.New(err.Error())
	}

	Duder.Rugs.VM.Set("print", func(msg string) { log.Println("[JS]", msg) })

	return nil
}

// response description
func response(value interface{}, defaultValue otto.Value) otto.Value {
	if v, err := Duder.Rugs.VM.ToValue(value); err == nil {
		return v
	}
	return defaultValue
}

/* Duder */
func rugenvSetStatus(call otto.FunctionCall) otto.Value {
	status := call.Argument(0).String()
	if len(status) == 0 {
		status = ""
	}
	Duder.Discord.SetStatus(status)
	Duder.Config.SetStatus(status)

	return otto.TrueValue()
}

func rugenvSetAvatar(call otto.FunctionCall) otto.Value {
	avatar := call.Argument(0).String()

	if ok := Duder.Discord.SetAvatarByImage(avatar); ok {
		return otto.TrueValue()
	}

	return otto.FalseValue()
}

func rugenvSaveAvatar(call otto.FunctionCall) otto.Value {
	filename := call.Argument(0).String()

	if err := Duder.Discord.SaveAvatar(filename); err != nil {
		return response(err.Error(), otto.FalseValue())
	}

	return otto.TrueValue()
}

func rugenvGetAvatars(call otto.FunctionCall) otto.Value {
	return response(Duder.Discord.Avatars(), otto.FalseValue())
}

func rugenvUseAvatar(call otto.FunctionCall) otto.Value {
	filename := call.Argument(0).String()

	if err := Duder.Discord.SetAvatarByFile(filename); err != nil {
		return response(err.Error(), otto.FalseValue())
	}

	return otto.TrueValue()
}

func rugenvStartTyping(call otto.FunctionCall) otto.Value {
	channelID := call.Argument(0).String()
	if len(channelID) == 0 {
		return otto.FalseValue()
	}

	ok := Duder.Discord.StartTyping(channelID)
	return response(ok, otto.FalseValue())
}

/* DuderUser */
func rugenvRugUserGetIsOwner(call otto.FunctionCall) otto.Value {
	userID := call.Argument(0).String()

	return response((userID == Duder.Config.OwnerID()), otto.FalseValue())
}

func rugenvRugUserGetIsManager(call otto.FunctionCall) otto.Value {
	guildID := call.Argument(0).String()
	userID := call.Argument(1).String()

	return response(Duder.Permissions.IsUserManager(guildID, userID), otto.FalseValue())
}

func rugenvRugUserGetIsModerator(call otto.FunctionCall) otto.Value {
	guildID := call.Argument(0).String()
	userID := call.Argument(1).String()

	return response(Duder.Permissions.IsUserModerator(guildID, userID), otto.FalseValue())
}

func rugenvRugUserSetNickname(call otto.FunctionCall) otto.Value {
	guildID := call.Argument(0).String()
	userID := call.Argument(1).String()
	nickname := call.Argument(1).String()

	Duder.Discord.SetMemberNickname(guildID, userID, nickname)

	return otto.TrueValue()
}

func rugenvRugUserGetUsernameByID(call otto.FunctionCall) otto.Value {
	guildID := call.Argument(0).String()
	userID := call.Argument(1).String()

	member, ok := Duder.Discord.GetGuildMember(guildID, userID)
	if !ok {
		return otto.FalseValue()
	}

	return response(member.User.Username, otto.FalseValue())
}

func rugenvRugUserGetIDByNickname(call otto.FunctionCall) otto.Value {
	guildID := call.Argument(0).String()
	nickname := call.Argument(1).String()

	member, ok := Duder.Discord.GetMemberByNickname(guildID, nickname)
	if !ok {
		return otto.FalseValue()
	}

	return response(member.User.ID, otto.FalseValue())
}

/* DuderCommand */
func rugenvRugCommandReplyToChannel(call otto.FunctionCall) otto.Value {
	channelID := call.Argument(0).String()
	content := call.Argument(1).String()

	//Duder.session.ChannelMessageSend(channelID, content)
	Duder.Discord.SendMessageToChannel(channelID, content)

	return otto.TrueValue()
}

func rugenvRugCommandReplyToChannelEmbed(call otto.FunctionCall) otto.Value {
	channelID := call.Argument(0).String()
	jsonData := call.Argument(1).String()

	if err := Duder.Discord.SendEmbedToChannel(channelID, jsonData); err != nil {
		Duder.Log(LogVerbose, "[cmd.replyToChannelEmbed] Failed to create embed", err.Error())
		return otto.FalseValue()
	}

	return otto.TrueValue()
}

func rugenvRugCommandReplyToAuthor(call otto.FunctionCall) otto.Value {
	channelID := call.Argument(0).String()
	authorID := call.Argument(1).String()
	authorUsername := call.Argument(2).String()
	content := call.Argument(3).String()
	mention, _ := call.Argument(4).ToBoolean()

	if mention {
		Duder.Discord.SendMessageToChannel(channelID, fmt.Sprintf("<@%s> %s", authorID, content))
	} else {
		Duder.Discord.SendMessageToChannel(channelID, fmt.Sprintf("%s, %s", authorUsername, content))
	}

	return otto.TrueValue()
}

func rugenvRugCommandDeleteMessage(call otto.FunctionCall) otto.Value {
	channelID := call.Argument(0).String()
	messageID := call.Argument(1).String()

	Duder.Discord.DeleteChannelMessage(channelID, messageID)

	return otto.TrueValue()
}

func rugenvRugCommandSendFile(call otto.FunctionCall) otto.Value {
	//channelID := call.Argument(0).String()
	//name := call.Argument(1).String()
	//data := call.Argument(2).String()

	return otto.TrueValue()
}

/* DuderRug */
func rugenvRugCreate(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	name := call.Argument(1).String()
	description := call.Argument(2).String()

	Duder.Rugs.CreateRug(rugObj, name, description)

	return otto.TrueValue()
}

func rugenvRugAddCommand(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	trigger := strings.TrimSpace(call.Argument(1).String())
	exec := call.Argument(2)

	// validate the trigger
	if len(trigger) == 0 {
		return otto.FalseValue()
	}

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.AddCommand(trigger, exec)
	}

	return otto.TrueValue()
}

func rugenvRugBindOnMessage(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	onMessage := call.Argument(1)

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.BindOnMessage(onMessage)
	}

	return otto.TrueValue()
}

func rugenvRugBindOnPresenceUpdate(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	onPresenceUpdate := call.Argument(1)

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.BindOnPresenceUpdate(onPresenceUpdate)
	}

	return otto.TrueValue()
}

func rugenvRugBindOnMessageReactionAdd(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	onMessageReactionAdd := call.Argument(1)

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.BindOnMessageReactionAdd(onMessageReactionAdd)
	}

	return otto.TrueValue()
}

func rugenvRugBindOnMessageReactionRemove(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	onMessageReactionRemove := call.Argument(1)

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.BindOnMessageReactionRemove(onMessageReactionRemove)
	}

	return otto.TrueValue()
}

func rugenvRugLoadStorage(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		storage, ok := rug.LoadStorage()
		if ok {
			return response(storage, otto.FalseValue())
		}

	}

	return otto.FalseValue()
}

func rugenvRugSaveStorage(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	data := call.Argument(1).String()

	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		return response(rug.SaveStorage(data), otto.FalseValue())
	}

	return otto.FalseValue()
}

func rugenvRugDPrint(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	msg := call.Argument(1).String()
	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.DPrint(msg)
		return otto.TrueValue()
	}
	return otto.FalseValue()
}

func rugenvRugWPrint(call otto.FunctionCall) otto.Value {
	rugObj := call.Argument(0).Object()
	msg := call.Argument(1).String()
	if rug, ok := Duder.Rugs.FindRugByObject(rugObj); ok {
		rug.WPrint(msg)
		return otto.TrueValue()
	}
	return otto.FalseValue()
}

/* String */
func rugenvStringDecodeHTML(call otto.FunctionCall) otto.Value {
	text := call.Argument(0).String()
	text = html.UnescapeString(text)

	return response(text, otto.FalseValue())
}

/* HTTP */
func rugenvHTTPGet(call otto.FunctionCall) otto.Value {
	var timeout int64
	timeout, _ = call.Argument(0).ToInteger()
	uri := call.Argument(1).String()
	headers, _ := call.Argument(2).Export()
	stringResult, _ := call.Argument(3).ToBoolean()

	client := http.Client{Timeout: time.Duration(timeout * int64(time.Second))}
	req, err := http.NewRequest("GET", uri, nil)
	if err != nil {
		return otto.FalseValue()
	}
	for k, v := range headers.(map[string]interface{}) {
		if value, ok := v.(string); ok {
			req.Header.Add(k, value)
			Duder.Logf(LogVerbose, "[HTTP.Get] Adding header '%s' : '%s'", k, value)
		}
	}

	resp, err := client.Do(req)
	if err != nil {
		Duder.Logf(LogVerbose, "[HTTP.Get] Error retrieving response; %s", err.Error())
		return otto.FalseValue()
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		Duder.Logf(LogVerbose, "[HTTP.Get] Error reading response body; %s", err.Error())
		return otto.FalseValue()
	}

	if stringResult {
		Duder.Log(LogVerbose, "[HTTP.Get] Returning string")
		return response(string(body), otto.FalseValue())
	}

	Duder.Log(LogVerbose, "[HTTP.Get] Returning byte array")
	return response(body, otto.FalseValue())
}

func rugenvHTTPPost(call otto.FunctionCall) otto.Value {
	var timeout int64
	timeout, _ = call.Argument(0).ToInteger()
	uri := call.Argument(1).String()
	values, _ := call.Argument(2).Export()

	client := http.Client{Timeout: time.Duration(timeout * int64(time.Second))}
	formValues := url.Values{}
	for k, v := range values.(map[string]interface{}) {
		if value, ok := v.(string); ok {
			formValues[k] = []string{value}
			Duder.Logf(LogVerbose, "[HTTP.Post] Adding form value '%s' : '%s'", k, value)
		}
	}

	resp, err := client.PostForm(uri, formValues)
	if err != nil {
		Duder.Logf(LogVerbose, "[HTTP.Post] Error posting form; %s", err.Error())
		return otto.FalseValue()
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		Duder.Logf(LogVerbose, "[HTTP.Post] Error reading response body; %s", err.Error())
		return otto.FalseValue()
	}

	return response(string(body), otto.FalseValue())
}

func rugenvHTTPDetectContentType(call otto.FunctionCall) otto.Value {
	var data []byte
	export, _ := call.Argument(0).Export()
	{
		data, _ = export.([]byte)
	}

	contentType := http.DetectContentType(data)
	if result, err := Duder.Rugs.VM.ToValue(contentType); err == nil {
		return result
	}
	return otto.FalseValue()
}

func rugenvHTTPParseURL(call otto.FunctionCall) otto.Value {
	urlString := call.Argument(0).String()
	if u, err := url.Parse(urlString); err == nil {
		if result, err := Duder.Rugs.VM.ToValue(u.String()); err == nil {
			return result
		}
	}
	return otto.FalseValue()
}

/* Base64 */
func rugenvBase64EncodeToString(call otto.FunctionCall) otto.Value {
	var data []byte
	export, _ := call.Argument(0).Export()
	{
		data, _ = export.([]byte)
	}

	str := base64.StdEncoding.EncodeToString(data)
	if result, err := Duder.Rugs.VM.ToValue(str); err == nil {
		return result
	}
	return otto.FalseValue()
}

/* XML */
func rugenvXMLtoJSON(call otto.FunctionCall) otto.Value {
	xml := call.Argument(0).String()
	str := strings.NewReader(xml)

	json, err := xml2json.Convert(str)
	if err != nil {
		return otto.FalseValue()
	}

	return response(json.String(), otto.FalseValue())
}

File: rugs_env.js
// Define Duder class
function Duder() {}
Duder.setStatus = function(status) {
	return __BIND__(status);
};
Duder.setAvatar = function(avatar) {
	return __BIND__(avatar);
};
Duder.saveAvatar = function(filename) {
	return __BIND__(filename);
};
Duder.getAvatars = function() {
	return __BIND__();
};
Duder.useAvatar = function(filename) {
	return __BIND__(filename);
};
Duder.startTyping = function(channelID) {
	return __BIND__(channelID);
};

// Define DuderUser class
function DuderUser(guildID, id, username) {
	this.id = id;
	this.username = username;
	this.isOwner = __BIND__(id);
	this.isManager = __BIND__(guildID, id);
	this.isModerator = __BIND__(guildID, id);
}
DuderUser.prototype.setNickname = function(guildID, nickname) {
	return __BIND__(guildID, this.id, nickname);
};
DuderUser.getUsernameByID = function(guildID, userID) {
	return __BIND__(guildID, userID);
};

// Define DuderCommand class
function DuderCommand(guildID, channelID, messageID, author, mentions, args) {
	this.guildID = guildID;
	this.channelID = channelID;
	this.messageID = messageID;
	this.author = author;
	this.mentions = mentions;
	this.args = args;
}
DuderCommand.prototype.replyToChannel = function(content) {
	__BIND__(this.channelID, content);
};
DuderCommand.prototype.replyToChannelEmbed = function(content) {
	__BIND__(this.channelID, content);
};
DuderCommand.prototype.replyToAuthor = function(content, mention) {
	// ensure 'mention' is bool and default is false
	mention = mention === true;
	__BIND__(
		this.channelID,
		this.author.id,
		this.author.username,
		content,
		mention
	);
};
DuderCommand.prototype.isMention = function(str) {
	return str.substring(0, 2) === "<@" && str.substring(str.length - 1) === ">";
};
DuderCommand.prototype.deleteMessage = function() {
	__BIND__(this.channelID, this.messageID);
};
DuderCommand.prototype.sendFile = function(channelID, name, data) {
	__BIND__(channelID, name, data);
};

// Define DuderRug class
function DuderRug(name, description) {
	__BIND__(this, name, description);
}
DuderRug.prototype.addCommand = function(trigger, exec) {
	__BIND__(this, trigger, exec);
};
DuderRug.prototype.loadStorage = function() {
	var data = __BIND__(this);
	return JSON.parse(data);
};
DuderRug.prototype.saveStorage = function(json) {
	var data = JSON.stringify(json, null, "\t");
	return __BIND__(this, data);
};
DuderRug.prototype.dprint = function(msg) {
	__BIND__(this, msg);
};
DuderRug.prototype.wprint = function(msg) {
	__BIND__(this, msg);
};

// Math
Math.getRandomInRange = function(min, max) {
	min = Math.ceil(min);
	max = Math.floor(max);
	return Math.floor(Math.random() * (max - min + 1)) + min;
};
Math.clamp = function(val, min, max) {
	return Math.max(min, Math.min(val, max));
};

// String
String.prototype.decodeHTML = function() {
	return __BIND__(this);
};
String.prototype.replaceAll = function(search, replacement) {
	var target = this;
	return target.replace(new RegExp(search, "g"), replacement);
};
String.prototype.toBinary = function() {
	var result = [];
	for (var i = 0; i < this.length; i++) {
		result.push(this.charCodeAt(i));
	}
	return result;
};
String.prototype.format = function() {
	var args = arguments;
	return this.replace(/{(\d+)}/g, function(match, number) {
		return typeof args[number] !== "undefined" ? args[number] : match;
	});
};
String.prototype.matchAll = function(regexp) {
	var matches = [];
	this.replace(regexp, function() {
		var arr = [].slice.call(arguments, 0);
		var extras = arr.splice(-2);
		arr.index = extras[0];
		arr.input = extras[1];
		matches.push(arr);
	});
	return matches.length ? matches : null;
};

// Array
Array.prototype.contains = function(elem) {
	for (var i in this) {
		if (this[i] === elem) return true;
	}
	return false;
};

// HTTP
function HTTP() {}
HTTP.get = function(timeout, url, headers, as_string) {
	// ensure 'as_string' is bool and default is true
	as_string = as_string !== false;

	// ensure 'headers' is an array
	if (headers === null || headers === undefined) {
		headers = {};
	} else if (!(headers instanceof Object)) {
		headers = {};
	}

	return __BIND__(timeout, url, headers, as_string);
};
HTTP.post = function(timeout, url, values) {
	return __BIND__(timeout, url, values);
};
HTTP.detectContentType = function(content) {
	return __BIND__(content);
};
HTTP.parseURL = function(str) {
	return __BIND__(str);
};

// Base64
function Base64() {}
Base64.encodeToString = function(bytes) {
	return __BIND__(bytes);
};

// Misc
function isNumeric(n) {
	return !isNaN(parseFloat(n)) && isFinite(n);
}

// EmbedMessage
function EmbedMessage() {
	this.data = {
		title: null,
		description: null,
		url: null,
		color: null,
		timestamp: null,
		footer: null,
		thumbnail: null,
		image: null,
		author: null,
		fields: null
	};
}
EmbedMessage.prototype.setTitle = function(title) {
	this.data.title = title;
};
EmbedMessage.prototype.setDescription = function(description) {
	this.data.description = description;
};
EmbedMessage.prototype.setURL = function(url) {
	this.data.url = url;
};
EmbedMessage.prototype.setColor = function(color) {
	this.data.color = color;
};
EmbedMessage.prototype.setTimestamp = function() {
	var dt = new Date();
	this.data.timestamp = dt.toISOString();
};
EmbedMessage.prototype.setFooter = function(icon, text) {
	this.data.footer = '{\n\t\t"icon_url": "{0}",\n\t\t"text": "{1}"\n\t}'.format(
		icon,
		text
	);
};
EmbedMessage.prototype.setThumbnail = function(thumbnail) {
	this.data.thumbnail = '{\n\t\t"url": "{0}"\n\t}'.format(thumbnail);
};
EmbedMessage.prototype.setImage = function(image) {
	this.data.image = '{\n\t\t"url": "{0}"\n\t}'.format(image);
};
EmbedMessage.prototype.setAuthor = function(name, url, icon) {
	this.data.author = '{\n\t\t"name": "{0}",\n\t\t"url": "{1}",\n\t\t"icon_url": "{2}"\n\t}'.format(
		name,
		url,
		icon
	);
};
EmbedMessage.prototype.addField = function(name, value) {
	if (this.data.fields === null) {
		this.data.fields = [];
	}
	var field = [name, value];
	this.data.fields.push(field);
};
EmbedMessage.prototype.compile = function() {
	var content = "";
	var COMMA = function(c) {
		return c.length > 0 ? ",\n" : "";
	};
	if (this.data.title !== null) {
		content += COMMA(content) + '\t"title": "{0}"'.format(this.data.title);
	}
	if (this.data.description !== null) {
		content +=
			COMMA(content) +
			'\t"description": "{0}"'.format(this.data.description);
	}
	if (this.data.url !== null) {
		content += COMMA(content) + '\t"url": "{0}"'.format(this.data.url);
	}
	if (this.data.color !== null) {
		content += COMMA(content) + '\t"color": {0}'.format(this.data.color);
	}
	if (this.data.timestamp !== null) {
		content +=
			COMMA(content) + '\t"timestamp": "{0}"'.format(this.data.timestamp);
	}
	if (this.data.footer !== null) {
		content += COMMA(content) + '\t"footer": {0}'.format(this.data.footer);
	}
	if (this.data.thumbnail !== null) {
		content +=
			COMMA(content) + '\t"thumbnail": {0}'.format(this.data.thumbnail);
	}
	if (this.data.image !== null) {
		content += COMMA(content) + '\t"image": {0}'.format(this.data.image);
	}
	if (this.data.author !== null) {
		content += COMMA(content) + '\t"author": {0}'.format(this.data.author);
	}
	if (this.data.fields !== null) {
		var fields = '\t"fields": [\n';
		for (var i = 0; i < this.data.fields.length; i++) {
			var field = this.data.fields[i];
			fields += '\t\t{\n\t\t\t"name": "{0}",\n\t\t\t"value": "{1}"\n\t\t}{2}\n'.format(
				field[0],
				field[1],
				i < this.data.fields.length - 1 ? "," : ""
			);
		}
		fields += "\t]";
		content += COMMA(content) + fields;
	}

	content = "{\n" + content + "\n}";

	return content;
};
